File: chapter_api_tests/2024-04/validation/test_put_api-v1-runs-{runId}-metadata.py
import axios, { AxiosError, AxiosResponse } from 'axios';
import dotenv from 'dotenv';

dotenv.config();

const API_BASE_URL = process.env.API_BASE_URL;
const API_AUTH_TOKEN = process.env.API_AUTH_TOKEN;

// Interface for a successful response
interface UpdateRunMetadataResponse {
  metadata: Record<string, any>;
}

// Interface for an error response
interface ErrorResponse {
  error: string;
}

describe('PUT /api/v1/runs/{runId}/metadata', () => {
  // Helper to create Axios instance with base config
  const createAxiosInstance = (token?: string) => {
    return axios.create({
      baseURL: API_BASE_URL,
      headers: {
        'Content-Type': 'application/json',
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
      validateStatus: () => true, // We will handle status checks in the tests
    });
  };

  const validRunId = '123';
  const nonExistentRunId = '999999';
  
  /**
   * Valid metadata payload to update
   */
  const validMetadataPayload = {
    metadata: {
      key1: 'value1',
      key2: 2,
    },
  };

  /**
   * Creates a very large metadata object for edge-case testing.
   */
  const largeMetadataPayload = {
    metadata: {
      largeKey: 'x'.repeat(10_000),
    },
  };

  describe('Input Validation & Response Validation', () => {
    test('should update run metadata with valid inputs (200)', async () => {
      const instance = createAxiosInstance(API_AUTH_TOKEN);

      const response: AxiosResponse<UpdateRunMetadataResponse> = await instance.put(
        `/api/v1/runs/${validRunId}/metadata`,
        validMetadataPayload
      );

      // Expecting successful update
      expect([200]).toContain(response.status);
      expect(response.data).toHaveProperty('metadata');
      expect(typeof response.data.metadata).toBe('object');

      // Header validation
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    });

    test('should return 400 or 422 when runId is invalid (string, empty, etc.)', async () => {
      const instance = createAxiosInstance(API_AUTH_TOKEN);
      // Trying a clearly invalid runId
      const invalidRunId = 'abc';

      const response: AxiosResponse<ErrorResponse> = await instance.put(
        `/api/v1/runs/${invalidRunId}/metadata`,
        validMetadataPayload
      );

      expect([400, 422]).toContain(response.status);
      // The response body might contain an "error" field with a relevant message.
      if (response.data) {
        expect(response.data).toHaveProperty('error');
      }

      // Header validation
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    });

    test('should return 400 or 422 if metadata payload is invalid (e.g., missing metadata)', async () => {
      const instance = createAxiosInstance(API_AUTH_TOKEN);
      // Invalid payload: missing or empty metadata field
      const invalidPayload = { badField: 'not metadata' };

      const response: AxiosResponse<ErrorResponse> = await instance.put(
        `/api/v1/runs/${validRunId}/metadata`,
        invalidPayload
      );

      expect([400, 422]).toContain(response.status);
      if (response.data) {
        expect(response.data).toHaveProperty('error');
      }

      // Header validation
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    });

    test('should return 404 if runId does not exist', async () => {
      const instance = createAxiosInstance(API_AUTH_TOKEN);

      const response: AxiosResponse<ErrorResponse> = await instance.put(
        `/api/v1/runs/${nonExistentRunId}/metadata`,
        validMetadataPayload
      );

      // Expecting 404 for run not found
      expect(response.status).toBe(404);
      if (response.data) {
        expect(response.data).toHaveProperty('error');
      }

      // Header validation
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    });
  });

  describe('Edge Case & Limit Testing', () => {
    test('should handle large payload without errors', async () => {
      const instance = createAxiosInstance(API_AUTH_TOKEN);

      const response: AxiosResponse<UpdateRunMetadataResponse | ErrorResponse> = await instance.put(
        `/api/v1/runs/${validRunId}/metadata`,
        largeMetadataPayload
      );

      // Some APIs might handle large payload differently, but assume success is 200
      expect([200, 400, 422]).toContain(response.status);
      // If success, ensure metadata is returned
      if (response.status === 200) {
        const successData = response.data as UpdateRunMetadataResponse;
        expect(successData).toHaveProperty('metadata');
      } else {
        // If not success, check we got an error structure
        const errorData = response.data as ErrorResponse;
        expect(errorData).toHaveProperty('error');
      }

      // Header validation
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    });

    test('should handle empty metadata as valid or invalid (depending on API spec)', async () => {
      const instance = createAxiosInstance(API_AUTH_TOKEN);

      // Some APIs accept an empty object, some do not.
      const emptyMetadataPayload = {
        metadata: {},
      };

      const response: AxiosResponse<UpdateRunMetadataResponse | ErrorResponse> = await instance.put(
        `/api/v1/runs/${validRunId}/metadata`,
        emptyMetadataPayload
      );

      // Expect success or an input error
      expect([200, 400, 422]).toContain(response.status);
      if (response.status === 200) {
        const successData = response.data as UpdateRunMetadataResponse;
        expect(successData).toHaveProperty('metadata');
      } else {
        const errorData = response.data as ErrorResponse;
        expect(errorData).toHaveProperty('error');
      }

      // Header validation
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    });
  });

  describe('Testing Authorization & Authentication', () => {
    test('should return 401 or 403 if API token is missing', async () => {
      const instance = createAxiosInstance(undefined); // no token

      const response: AxiosResponse<ErrorResponse> = await instance.put(
        `/api/v1/runs/${validRunId}/metadata`,
        validMetadataPayload
      );

      expect([401, 403]).toContain(response.status);
      if (response.status === 401 || response.status === 403) {
        expect(response.data).toHaveProperty('error');
      }

      // Header validation
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    });

    test('should return 401 or 403 if API token is invalid', async () => {
      const instance = createAxiosInstance('invalid-token');

      const response: AxiosResponse<ErrorResponse> = await instance.put(
        `/api/v1/runs/${validRunId}/metadata`,
        validMetadataPayload
      );

      expect([401, 403]).toContain(response.status);
      if (response.status === 401 || response.status === 403) {
        expect(response.data).toHaveProperty('error');
      }

      // Header validation
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    });
  });

  // Additional tests could cover server errors or other edge cases (e.g., 500 responses).
});

File: chapter_api_tests/2024-04/validation/test_put_api-v1-runs-{runId}-metadata.ts
import axios, { AxiosInstance, AxiosResponse } from 'axios';

describe('PUT /api/v1/runs/:runId/metadata', () => {
  let client: AxiosInstance;
  // Adjust these IDs based on your actual data/environment
  const validRunId = '12345'; // Replace with a known valid run ID if available
  const invalidRunId = 'abc'; // Example of an invalid run ID
  const nonExistentRunId = '99999'; // Example of a run ID that doesn't exist

  beforeAll(() => {
    const baseURL = process.env.API_BASE_URL || '';
    const token = process.env.API_AUTH_TOKEN || '';

    // Create an Axios instance with baseURL and authorization header
    client = axios.create({
      baseURL,
      headers: {
        'Authorization': `Bearer ${token}`,
      },
      // validateStatus allows us to receive non-2xx responses without throwing
      validateStatus: () => true,
    });
  });

  it('should update run metadata with a valid payload (200)', async () => {
    const payload = {
      metadata: { exampleKey: 'exampleValue' },
    };

    // Make the PUT request with a valid runId and valid payload
    const response: AxiosResponse = await client.put(`/api/v1/runs/${validRunId}/metadata`, payload);

    // Check for expected 200 response
    expect(response.status).toBe(200);
    // Validate response headers
    expect(response.headers['content-type']).toMatch(/application\\/json/);
    // Validate response body schema
    expect(response.data).toHaveProperty('metadata');
    expect(typeof response.data.metadata).toBe('object');
  });

  it('should return 401 or 403 for unauthorized or forbidden request', async () => {
    // Create a client with no/invalid token
    const unauthorizedClient = axios.create({
      baseURL: process.env.API_BASE_URL || '',
      validateStatus: () => true,
    });

    const payload = {
      metadata: { exampleKey: 'exampleValue' },
    };

    const response: AxiosResponse = await unauthorizedClient.put(`/api/v1/runs/${validRunId}/metadata`, payload);

    // The API might return 401 or 403
    expect([401, 403]).toContain(response.status);
    expect(response.headers['content-type']).toMatch(/application\\/json/);
    expect(response.data).toHaveProperty('error');
  });

  it('should return 400 or 422 for invalid payload (e.g., metadata is not an object)', async () => {
    const payload = {
      metadata: 'this_should_be_an_object', // Invalid type
    };

    const response: AxiosResponse = await client.put(`/api/v1/runs/${validRunId}/metadata`, payload);

    // The API may return 400 or 422
    expect([400, 422]).toContain(response.status);
    expect(response.headers['content-type']).toMatch(/application\\/json/);
    expect(response.data).toHaveProperty('error');
  });

  it('should return 400 or 422 when the request body is empty', async () => {
    const response: AxiosResponse = await client.put(`/api/v1/runs/${validRunId}/metadata`, {});

    // The API may return 400 or 422
    expect([400, 422]).toContain(response.status);
    expect(response.headers['content-type']).toMatch(/application\\/json/);
    expect(response.data).toHaveProperty('error');
  });

  it('should return 400 if runId is invalid format', async () => {
    const payload = {
      metadata: { exampleKey: 'exampleValue' },
    };

    const response: AxiosResponse = await client.put(`/api/v1/runs/${invalidRunId}/metadata`, payload);

    // Depending on the API, it might return 400 or 404
    expect([400, 404]).toContain(response.status);
    if (response.status === 400) {
      // Validate error response
      expect(response.headers['content-type']).toMatch(/application\\/json/);
      expect(response.data).toHaveProperty('error');
      // Check one of the allowed error messages
      expect(response.data.error).toMatch(/Invalid or missing run ID|Invalid metadata/);
    }
    if (response.status === 404) {
      expect(response.headers['content-type']).toMatch(/application\\/json/);
      expect(response.data).toHaveProperty('error');
      expect(response.data.error).toBe('Task Run not found');
    }
  });

  it('should return 404 if run does not exist', async () => {
    const payload = {
      metadata: { exampleKey: 'exampleValue' },
    };

    const response: AxiosResponse = await client.put(`/api/v1/runs/${nonExistentRunId}/metadata`, payload);

    // Some APIs may return 400 if runId is invalid, or 404 if not found
    expect([400, 404]).toContain(response.status);
    if (response.status === 404) {
      expect(response.headers['content-type']).toMatch(/application\\/json/);
      expect(response.data).toHaveProperty('error');
      expect(response.data.error).toBe('Task Run not found');
    }
  });

  it('should handle large payload gracefully', async () => {
    // Create a large metadata object to test boundary/limit scenarios
    const largeMetadata: Record<string, string> = {};
    for (let i = 0; i < 10000; i++) {
      largeMetadata[`key${i}`] = `value${i}`;
    }

    const payload = { metadata: largeMetadata };

    const response: AxiosResponse = await client.put(`/api/v1/runs/${validRunId}/metadata`, payload);

    // Could be 200 if successful, 400/413 if too large, etc.
    expect([200, 400, 413]).toContain(response.status);
    expect(response.headers['content-type']).toMatch(/application\\/json/);
  });

  it('should return 401 if missing authorization header', async () => {
    // Create a client with no auth header
    const noAuthClient = axios.create({
      baseURL: process.env.API_BASE_URL || '',
      validateStatus: () => true,
    });

    const payload = {
      metadata: { exampleKey: 'exampleValue' },
    };

    const response: AxiosResponse = await noAuthClient.put(`/api/v1/runs/${validRunId}/metadata`, payload);

    // The API might return 401 or 403
    expect([401, 403]).toContain(response.status);
    expect(response.headers['content-type']).toMatch(/application\\/json/);
    expect(response.data).toHaveProperty('error');
    expect(response.data.error).toMatch(/Invalid or Missing API key/);
  });

  it('should handle unexpected server errors (500) gracefully', async () => {
    // There's no guaranteed way to trigger a 500, but we can try unusual payloads
    const payload = {
      metadata: { triggerServerError: true },
    };

    const response: AxiosResponse = await client.put(`/api/v1/runs/${validRunId}/metadata`, payload);

    // If 500 occurs, validate the error body
    if (response.status === 500) {
      expect(response.headers['content-type']).toMatch(/application\\/json/);
      expect(response.data).toHaveProperty('error');
    } else {
      // Otherwise we expect some valid status like 200, 400, or 422
      expect([200, 400, 422].includes(response.status)).toBe(true);
    }
  });
});